src/widget.rs

```diff
@@ -30,57 +30,184 @@ use std::ops::{Deref, DerefMut};
+    ///
+    /// [`PaintCtx`]: struct.PaintCtx.html
+    /// [`RenderContext`]: trait.RenderContext.html
+    fn paint(&mut self, cx: &mut PaintCx, builder: &mut SceneBuilder);
+
+    /*
+    #[doc(hidden)]
+    /// Get the identity of the widget; this is basically only implemented by
+    /// `IdentityWrapper`. Widgets should not implement this on their own.
+    fn id(&self) -> Option<WidgetId> {
+        None
+    }
+    */
+
+    #[doc(hidden)]
+    /// Get the (verbose) type name of the widget for debugging purposes.
+    /// You should not override this method.
+    fn type_name(&self) -> &'static str {
+        std::any::type_name::<Self>()
+    }
+
+    #[doc(hidden)]
+    /// Get the (abridged) type name of the widget for debugging purposes.
+    /// You should not override this method.
+    fn short_type_name(&self) -> &'static str {
+        let name = self.type_name();
+        name.split('<')
+            .next()
+            .unwrap_or(name)
+            .split("::")
+            .last()
+            .unwrap_or(name)
+    }
+
+    /*
+    #[doc(hidden)]
+    /// From the current data, get a best-effort description of the state of
+    /// this widget and its children for debugging purposes.
+    fn debug_state(&self) -> DebugState {
+        #![allow(unused_variables)]
+        DebugState {
+            display_name: self.short_type_name().to_string(),
+            ..Default::default()
+        }
+    }
+    */
 
-    /// Prepare for painting.
+    /// Computes max intrinsic/preferred dimension of a widget on the provided axis.
+    ///
+    /// Max intrinsic/preferred dimension is the dimension the widget could take, provided infinite
+    /// constraint on that axis.
+    ///
+    /// If axis == Axis::Horizontal, widget is being asked to calculate max intrinsic width.
+    /// If axis == Axis::Vertical, widget is being asked to calculate max intrinsic height.
+    ///
+    /// Box constraints must be honored in intrinsics computation.
     ///
-    /// This method is currently a bit of a hack. It's similar to the one in
-    /// Druid, which is for incremental repaint, but the primary purpose of
-    /// this one is virtualized scrolling.
+    /// AspectRatioBox is an example where constraints are honored. If height is finite, max intrinsic
+    /// width is *height * ratio*.
+    /// Only when height is infinite, child's max intrinsic width is calculated.
     ///
-    /// The fact that `cx` is LayoutCx is just laziness, it should have its
-    /// own cx. And the main methods on that cx should be for region-based
-    /// invalidation.
-    #[allow(unused)]
-    fn prepare_paint(&mut self, cx: &mut LayoutCx, visible: Rect) {}
+    /// Intrinsic is a *could-be* value. It's the value a widget *could* have given infinite constraints.
+    /// This does not mean the value returned by layout() would be the same.
+    ///
+    /// This method **must** return a finite value.
+    fn compute_max_intrinsic(
```

Comment:
I think we should either add a `compute_max_intrinsic` method to `Pod` or remove this one.

Reply:
I agree, and I also think that should happen in a followup PR. The scope of this one is focused on the `Widget` trait itself, and it's known that there's a bunch more wiring, especially in `Pod`, that needs to be done.